### DISCUSSION TITLE
mise failed to install "cargo:just" = "1.37.0"

### DISCUSSION THREAD

**Discussion by zhiqiangxu** [jdx/mise#3663] [Troubleshooting and bug reports]
**Describe the bug**
When I run `mise install --verbose` in the dir with [mise.toml](https://github.com/ethereum-optimism/optimism/blob/develop/mise.toml), it fails with this error:
<img width="827" alt="image" src="https://github.com/user-attachments/assets/ba31b92e-9113-49e7-8fb8-8ae9b2178054" />

I think the root cause is that mise doesn't handle `linker `cc` not found` correctly, as it's gone after I run `apt install gcc` manually.

**To Reproduce**
git clone https://github.com/ethereum-optimism/optimism
cd optimism
mise trust
mise install --verbose

**Expected behavior**
All tools are installed successfully.

**`mise doctor` output**

```text
# mise doctor
version: 2024.12.13 linux-x64 (c970664 2024-12-17)
activated: yes
shims_on_path: no

build_info:
  Target: x86_64-unknown-linux-gnu
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Tue, 17 Dec 2024 22:05:20 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26)
  Profile: release

shell:
  /bin/bash
  GNU bash, version 5.2.21(1)-release (x86_64-pc-linux-gnu)
  Copyright (C) 2022 Free Software Foundation, Inc.
  License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

  This is free software; you are free to change and redistribute it.
  There is NO WARRANTY, to the extent permitted by law.

dirs:
  cache: ~/.cache/mise
  config: ~/.config/mise
  data: ~/.local/share/mise
  shims: ~/.local/share/mise/shims
  state: ~/.local/state/mise

config_files:
  ~/xu/local_devnet/optimism/mise.toml

backends:
  aqua
  asdf
  cargo
  core
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:

toolset:
  aqua:astral-sh/uv@0.5.5
  aqua:direnv/direnv@2.35.0
  aqua:jqlang/jq@1.7.1
  aqua:koalaman/shellcheck@0.10.0
  cargo:just@1.37.0     (missing)
  cargo:svm-rs@0.5.8    (missing)
  core:go@1.22.7
  core:python@3.12.0
  core:rust@1.83.0
  go:github.com/ethereum/go-ethereum/cmd/abigen@1.10.25
  go:github.com/ethereum/go-ethereum/cmd/geth@1.14.7
  go:github.com/golangci/golangci-lint/cmd/golangci-lint@1.61.0
  go:github.com/mikefarah/yq/v4@4.44.3
  go:github.com/protolambda/eth2-testnet-genesis@0.10.0
  go:github.com/vektra/mockery/v2@2.46.0
  go:gotest.tools/gotestsum@1.12.0
  pipx:semgrep@1.90.0
  pipx:slither-analyzer@0.10.2
  ubi:casey/just@1.37.0
  ubi:foundry-rs/foundry[exe=anvil]@nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
  ubi:foundry-rs/foundry[exe=cast]@nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
  ubi:foundry-rs/foundry[exe=forge]@nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b
  ubi:mikefarah/yq@4.44.5

path:
  ~/.local/share/mise/installs/go/1.22.7/bin
  ~/.cargo/bin
  ~/.local/share/mise/installs/python/3.12.0/bin
  ~/.local/share/mise/installs/uv/0.5.5/uv-x86_64-unknown-linux-musl
  ~/.local/share/mise/installs/jq/1.7.1
  ~/.local/share/mise/installs/yq/4.44.5/bin
  ~/.local/share/mise/installs/shellcheck/0.10.0/shellcheck-v0.10.0
  ~/.local/share/mise/installs/direnv/2.35.0
  ~/.local/share/mise/installs/just/1.37.0/bin
  ~/.local/share/mise/installs/go-github-com-ethereum-go-ethereum-cmd-abigen/1.10.25/bin
  ~/.local/share/mise/installs/go-github-com-ethereum-go-ethereum-cmd-geth/1.14.7/bin
  ~/.local/share/mise/installs/go-github-com-protolambda-eth2-testnet-genesis/0.10.0/bin
  ~/.local/share/mise/installs/go-gotest-tools-gotestsum/1.12.0/bin
  ~/.local/share/mise/installs/go-github-com-vektra-mockery-v2/2.46.0/bin
  ~/.local/share/mise/installs/go-github-com-golangci-golangci-lint-cmd-golangci-lint/1.61.0/bin
  ~/.local/share/mise/installs/go-github-com-mikefarah-yq-v4/4.44.3/bin
  ~/.local/share/mise/installs/pipx-slither-analyzer/0.10.2/bin
  ~/.local/share/mise/installs/pipx-semgrep/1.90.0/bin
  ~/.local/share/mise/installs/forge/nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/bin
  ~/.local/share/mise/installs/cast/nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/bin
  ~/.local/share/mise/installs/anvil/nightly-e5dbb7a320c2b871c4a4a1006ad3c15a08fcf17b/bin
  ~/.local/bin
  /usr/local/sbin
  /usr/local/bin
  /usr/sbin
  /usr/bin
  /sbin
  /bin
  /usr/games
  /usr/local/games
  /snap/bin

env_vars:
  MISE_SHELL=bash

settings:
  disable_tools  ["asterisc", "binary_signer", "kontrol"] ~/xu/local_devnet/optimism/mise.toml
  experimental   true                                     ~/xu/local_devnet/optimism/mise.toml
  pipx.uvx       true                                     ~/xu/local_devnet/optimism/mise.toml

1 problem found:

1. shims are missing, run mise reshim to create them
   Missing shims: 2to3, 2to3-3.12, abigen, anvil, cargo, cargo-clippy, cargo-fmt, cargo-miri, cast, clippy-driver, direnv, eth2-testnet-genesis, forge, geth, go, gofmt, golangci-lint, gotestsum, idle3, idle3.12, jq, just, mockery, pip, pip3, pip3.12, pydoc3, pydoc3.12, pysemgrep, python, python3, python3-config, python3.12, python3.12-config, rls, rust-analyzer, rust-gdb, rust-gdbgui, rust-lldb, rustc, rustdoc, rustfmt, rustup, semgrep, shellcheck, slither, slither-check-erc, slither-check-kspec, slither-check-upgradeability, slither-doctor, slither-documentation, slither-find-paths, slither-flat, slither-format, slither-interface, slither-mutate, slither-prop, slither-read-storage, slither-simil, uv, uvx, yq

```

**Additional context**
N/A

**Comment by syhol** [comment]
This is expected. Mise isn't a dependency manager, so it won't try to install things like libraries or c compilers for you: https://mise.jdx.dev/roadmap.html#anti-goals this is why static binaries are preferred and you'll notice they have higher priority in the registry, e.g. `just` maps first to ubi:casey/just.

Also you might not need `just` if you use [mise tasks](https://mise.jdx.dev/tasks/).

**Comment by zhiqiangxu** [comment]
> Mise isn't a dependency manager

What would you call it, tool manager? In general, any tool is also a dependency though.

I read the roadmap link and it says:

> While this seems like an obvious downside, it actually ends up making mise far easier to use than nix. 

Could you provide an example showing mise is far easier to use than nix?

**Reply by JamBalaya56562** [reply]
Not a maintainer, just chiming in 👋

**On terminology:** the line mise draws is between **developer tools** and **system/build dependencies**. mise manages tool/runtime *versions* (node, python, go, rust, CLIs) plus env vars and tasks — the asdf-style "dev tool version manager" role. What it deliberately won't do is install **system libraries / build toolchains** (a C compiler, OpenSSL headers, glibc, …); that's what "not a dependency manager" means. So yes, a tool is loosely a "dependency," but the anti-goal is specifically about not reimplementing an OS package manager for system libs. That's exactly your original case: `cargo:just` is compiled from source and needed a C linker (`cc`), which is out of scope — hence `apt install gcc` fixed it, or you can use the prebuilt `ubi:casey/just` and skip compiling entirely.

**mise vs nix — a concrete example.** Say a project needs Node 20, Python 3.12, and Terraform 1.9:

mise:
```toml
# mise.toml
[tools]
node = "20"
python = "3.12"
terraform = "1.9"
```
`mise install`, and they're on `PATH`; entering/leaving the dir auto-activates them. A familiar declarative file, no new language to learn.

With nix you'd write a `flake.nix`/`shell.nix` in the Nix language and `nix develop`. The friction is pinning *specific* versions: a given `nixpkgs` revision generally ships one version of each package, so getting an arbitrary older Terraform usually means overlays or hunting down a matching nixpkgs commit (tools like lazamar's nix-versions exist precisely because that's awkward).

The tradeoff is real, though: nix buys far stronger reproducibility and can manage system libraries too. mise intentionally scopes down to tool versions + env + tasks, which is why it's quicker to pick up for the common "I just need these tool versions" case. Different tools for different needs.

*This comment was generated by an AI coding assistant.*
